src/main/java/org/javarosa/xpath/expr/XPathSortFunc.java

```diff
@@ -0,0 +1,146 @@
+        List<String> items = stringToList(spaceSeparatedString);
+        sortSingleList(items, ascending);
+        return items;
+    }
+
+    private void sortSingleList(List<String> items, final boolean ascending) {
+        Collections.sort(items, new Comparator<String>() {
+            @Override
+            public int compare(String s1, String s2) {
+                return (ascending ? 1 : -1) * s1.compareTo(s2);
+            }
+        });
+    }
+
+    private List<String> sortListByOtherList(String s1, String s2, boolean ascending) {
+        List<String> targetListItems = stringToList(s1);
+        List<String> comparisonListItems = stringToList(s2);
+
+        if (targetListItems.size() != comparisonListItems.size()) {
+            throw new XPathTypeMismatchException("Length of lists passed to sort() must match");
+        }
+
+        Map<String, List<String>> stringMapping =
+                createMappingFromComparisonToTarget(comparisonListItems, targetListItems);
+
+        List<String> sortedComparisonList = sortSingleList(s2, ascending);
+        List<String> sortedTargetList = new ArrayList<>();
+        String previousComparisonString = "";
+        for (int i = 0; i < sortedComparisonList.size(); i++) {
+            String stringInSortedList = sortedComparisonList.get(i);
+            if (stringInSortedList.equals(previousComparisonString)) {
+                // Means we already grabbed all the target strings corresponding to this reference string
+                continue;
+            }
+            List<String> correspondingStrings = stringMapping.get(stringInSortedList);
+            if (correspondingStrings.size() > 1) {
+                sortSingleList(correspondingStrings, ascending);
+            }
+            sortedTargetList.addAll(correspondingStrings);
+            previousComparisonString = stringInSortedList;
+        }
+
+        return sortedTargetList;
+    }
+
+    private static Map<String, List<String>> createMappingFromComparisonToTarget(List<String> comparisonListItems,
+                                                                           List<String> targetListItems) {
+        Map<String, List<String>> stringMapping = new HashMap<>();
+        for (int i = 0; i < comparisonListItems.size(); i++) {
+            String comparisonString = comparisonListItems.get(i);
+            String targetString = targetListItems.get(i);
+            List<String> correspondingStrings;
+            if (stringMapping.containsKey(comparisonString)) {
+                correspondingStrings = stringMapping.get(comparisonString);
+            } else {
+                correspondingStrings = new ArrayList<>();
+                stringMapping.put(comparisonString, correspondingStrings);
+            }
+            correspondingStrings.add(targetString);
+        }
+        return stringMapping;
+    }
+
+    private static List<String> stringToList(String s) {
```

Comment:
I looked for these and didn't find them (including in either of our `StringUtils` classes or in `ArrayUtilities`). If you know where they are, let me know and I'll switch over